File: Assignment/Chapter 1/Student Grade Calculator/Dart Version/lib/core/contracts/report_exporter.dart
import '../models/export_artifact.dart';
import '../models/export_format.dart';
import '../models/processing_report.dart';

abstract interface class ReportExporter {
  ExportFormat get format;

  Future<ExportArtifact> export(
    ProcessingReport report,
    String destinationPath,
  );
}

File: Assignment/Chapter 1/Student Grade Calculator/Kotlin Version/app/src/main/java/com/romualdsigning/studentgradecalc/domain/service/ReportExporter.kt
package com.romualdsigning.studentgradecalc.domain.service

import android.content.Context
import android.net.Uri
import com.romualdsigning.studentgradecalc.domain.model.ExportArtifact
import com.romualdsigning.studentgradecalc.domain.model.ExportDestination
import com.romualdsigning.studentgradecalc.domain.model.ExportFormat
import com.romualdsigning.studentgradecalc.domain.model.ProcessingReport

interface ReportExporter {
    val format: ExportFormat

    suspend fun export(
        context: Context,
        report: ProcessingReport,
        destination: Uri,
        exportDestination: ExportDestination,
    ): ExportArtifact
}
